src/Gaffer/ScriptNode.cpp

```diff
@@ -534,48 +534,42 @@ void ScriptNode::paste( Node *parent )
 	}
 }
 
+namespace
+{
+
 void handleReconnectionOfOutputPlug( const ScriptNode *script, const DependencyNode *dependencyNode, const Plug *outputPlug )
 {
-	// for plugs that hold output plugs we need to recurse
-	if(outputPlug->typeId() == Plug::staticTypeId())
+	const Plug *inPlug = dependencyNode->correspondingInput( outputPlug );
+	if ( inPlug )
 	{
-		for( OutputPlugIterator it( outputPlug ); !it.done(); ++it )
+		const Plug *srcPlug = inPlug->getInput<Plug>();
+		if ( srcPlug )
 		{
-			handleReconnectionOfOutputPlug( script, dependencyNode, it->get() );
+			// record this plug's current outputs, and reconnect them. This is a copy of outputPlug->outputs() rather
+			// than a reference, as reconnection can modify outputPlug->outputs()...
+			Plug::OutputContainer outputs = outputPlug->outputs();
+			for ( Plug::OutputContainer::const_iterator oIt = outputs.begin(); oIt != outputs.end(); ++oIt )
+			{
+				Plug *dstPlug = *oIt;
+				if ( dstPlug && dstPlug->acceptsInput( srcPlug ) && script->isAncestorOf( dstPlug ) )
+				{
+					dstPlug->setInput( const_cast<Gaffer::Plug*>(srcPlug) );
+					return;
```

Comment:
This return is too early, because it means one we've reconnected the first output for this plug, we'll fail to reconnect any others. In fact I was wrong to suggest that we should ever return early - even if an output exists at the parent level, it doesn't mean there couldn't be independent outputs at lower levels. This brings me to the embarrassing realisation that this entire change can be achieved by switching the original iterator to a recursive one (I've included the ++oIt changes too, because they're worth keeping) :

```
diff --git a/src/Gaffer/ScriptNode.cpp b/src/Gaffer/ScriptNode.cpp
index 625bdae..7964214 100644
--- a/src/Gaffer/ScriptNode.cpp
+++ b/src/Gaffer/ScriptNode.cpp
@@ -549,7 +549,7 @@ void ScriptNode::deleteNodes( Node *parent, const Set *filter, bool reconnect )
                        DependencyNode *dependencyNode = IECore::runTimeCast<DependencyNode>( node );
                        if( reconnect && dependencyNode )
                        {
-                               for( OutputPlugIterator it( node ); !it.done(); ++it )
+                               for( RecursiveOutputPlugIterator it( node ); !it.done(); ++it )
                                {
                                        Plug *inPlug = dependencyNode->correspondingInput( it->get() );
                                        if ( !inPlug )
@@ -566,18 +566,13 @@ void ScriptNode::deleteNodes( Node *parent, const Set *filter, bool reconnect )
                                        // record this plug's current outputs, and reconnect them. This is a copy of (*it)->outputs() rather
                                        // than a reference, as reconnection can modify (*it)->outputs()...
                                        Plug::OutputContainer outputs = (*it)->outputs();
-                                       for ( Plug::OutputContainer::const_iterator oIt = outputs.begin(); oIt != outputs.end(); )
+                                       for( Plug::OutputContainer::const_iterator oIt = outputs.begin(); oIt != outputs.end(); ++oIt )
                                        {
                                                Plug *dstPlug = *oIt;
                                                if ( dstPlug && dstPlug->acceptsInput( srcPlug ) && this->isAncestorOf( dstPlug ) )
                                                {
-                                                       oIt++;
                                                        dstPlug->setInput( srcPlug );
                                                }
-                                               else
-                                               {
-                                                       oIt++;
-                                               }
                                        }
                                }
                        }
```

Sorry for not spotting this sooner. I've run your new ScriptNode tests with this new version and they pass, so could you update your PR to use the approach above?